core/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -1259,28 +1260,40 @@ public static String digestToString(long digest) {
         private long writtenBytes;
         private final long checksumPosition;
         private String actualChecksum;
+        private final byte[] footerChecksum = new byte[8]; // this holds the actual footer checksum data written by to this output
 
         LuceneVerifyingIndexOutput(StoreFileMetaData metadata, IndexOutput out) {
             super(out);
             this.metadata = metadata;
-            checksumPosition = metadata.length() - 8; // the last 8 bytes are the checksum
+            checksumPosition = metadata.length() - 8; // the last 8 bytes are the checksum - we store it in footerChecksum
         }
 
         @Override
         public void verify() throws IOException {
+            String footerChecksum = null;
             if (metadata.checksum().equals(actualChecksum) && writtenBytes == metadata.length()) {
-                return;
+                ByteArrayIndexInput indexInput = new ByteArrayIndexInput("checksum", this.footerChecksum);
+                footerChecksum = digestToString(indexInput.readLong());
+                if (metadata.checksum().equals(footerChecksum)) {
+                    return;
+                }
             }
             throw new CorruptIndexException("verification failed (hardware problem?) : expected=" + metadata.checksum() +
-                    " actual=" + actualChecksum + " writtenLength=" + writtenBytes + " expectedLength=" + metadata.length() +
+                    " actual=" + actualChecksum + " footer=" + footerChecksum +" writtenLength=" + writtenBytes + " expectedLength=" + metadata.length() +
                     " (resource=" + metadata.toString() + ")", "VerifyingIndexOutput(" + metadata.name() + ")");
         }
 
         @Override
         public void writeByte(byte b) throws IOException {
-            if (writtenBytes++ == checksumPosition) {
+            if (writtenBytes == checksumPosition) {
                 readAndCompareChecksum();
+            } else if (writtenBytes > checksumPosition) { // we are writing parts of the checksum....
+                final long index = writtenBytes - checksumPosition;
+                if (index < footerChecksum.length) {
+                    footerChecksum[(int)index] = b;
+                }
```

Comment:
Can we throw an exc in an else clause here?  I.e., we should never try to write beyond the checksum...?